papis/zenodo.py

```diff
@@ -0,0 +1,221 @@
+    new_data["file_urls"] = data["files"]
+
+    return new_data
+
+
+def is_valid_record_id(record_id: str) -> bool:
+    """Checks if a record is a valid Zenodo record, first checking its form and then
+    testing against Zenodo
+
+    :param record_id: a Zenodo record id
+    :type record_id: str
+    :return: whether the record is valid
+    :rtype: bool
+    """
+    record_id = record_id.strip()
+    if not record_id.isdigit():
+        return False
+
+    with papis.utils.get_session() as session:
+        response = session.get(ZENODO_URL.format(record_id=record_id))
+
+    return response.ok
+
+
+def get_data(record_id: str) -> Dict[str, Any]:
+    """Fetches a record from the Zenodo API and processes it with a helper function
+
+    :param record_id: a Zenodo record id
+    :type record_id: str
+    :return: a processed zenodo record
+    :rtype: Dict[str, Any]
+    """
+    with papis.utils.get_session() as session:
+        response = session.get(
+            ZENODO_URL.format(record_id=record_id.strip()),
+            headers={"user-agent": f"papis/{papis.__version__}"},
+        )
+
+    try:
+        json_data = json.loads(response.content.decode())
+    except json.JSONDecodeError as exc:
+        logger.error("Failed to decode response from Zenodo.", exc_info=exc)
+
+    return zenodo_data_to_papis_data(json_data)
+
+
+class LinkInfo(TypedDict):
+    """Represents a the url for a Zenodo record.
+    Attributes:
+    - id, the uuid of the file
+    """
+
+    self: str
+
+
+class FileInfo(TypedDict):
+    """Represents a file info entry from a Zenodo record.
+
+    - id, the uuid of the file
+    - key, the filename
+    - size: the size of a file in bytes
+    - checksum: md5sum hash of the file.
+    - links: dict with the url to download the file
+    """
+
+    id: str
+    key: str
+    size: int
+    checksum: str
+    links: LinkInfo
+
+
+class Importer(papis.importer.Importer):
+    """Importer downloading data from a Zenodo ID"""
+
+    def __init__(self, uri: str = "") -> None:
+        super().__init__(name="zenodo", uri=uri)
+        self._file_urls: List[FileInfo] = []
+
+    @classmethod
+    def match(cls, uri: str) -> Optional[papis.importer.Importer]:
+        if is_valid_record_id(uri):
+            return Importer(uri)
+
+        return None
+
+    @classmethod
+    def match_data(cls, data: Dict[str, Any]) -> Optional["Importer"]:
+        return None
+
+    def fetch_data(self) -> None:
+        self.ctx.data = get_data(self.uri)
+        self._file_urls = self.ctx.data["file_urls"]
+        del self.ctx.data["file_urls"]
+
+    def fetch_files(self) -> None:
+        if not self.ctx.data:
+            return
+
+        for url_data in self._file_urls:
+            url = url_data["links"]["self"]
+            self.logger.info("Trying to download document from '%s'.", url)
+
+            filename = download_document(url, filename=url_data["key"])
```

Comment:
These files will be renamed in `papis add` or `papis update` using `add-file-name` anyway, so I'm not sure it's worth giving them a fixed file name here.

Most of our importers just get the one file, so it's not confusing. I imagine Zenodo downloads more files, so we might need to rework that..